traffic_router/connector/src/main/java/com/comcast/cdn/traffic_control/traffic_router/secure/CertificateDataConverter.java

```diff
@@ -17,30 +17,63 @@
 
 import com.comcast.cdn.traffic_control.traffic_router.shared.CertificateData;
 import org.apache.log4j.Logger;
+import org.bouncycastle.jcajce.provider.asymmetric.rsa.BCRSAPrivateCrtKey;
+import sun.security.rsa.RSAPrivateCrtKeyImpl;
+import sun.security.rsa.RSAPublicKeyImpl;
 
+import java.math.BigInteger;
 import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
 import java.security.cert.X509Certificate;
+import java.util.ArrayList;
 import java.util.List;
```

Comment:
Now, I'm not a big Java guy, so this could be very wrong, but couldn't you just use a `List` instead of an `ArrayList`? From what I understand, that leaves the list implementation up to the compiler, which in general allows better optimization. (it's also one less import)

Reply:
That's not how Java works. `List` is an interface of which you need a concrete implementation for, hence the `ArrayList`. Declaring it as an interface and assigning it a concrete implementation is generally best practice, as it allows you to swap in a different concrete implementation more easily without having to change every single place the `List` is used. For example, if we decided a `LinkedList` would work better for some reason, we'd only have to change it on this one line:
```
final List<X509Certificate> x509Chain = new ArrayList<>();
```

Reply:
Ah, I see.